src/puffin/src/puffin_manager/cache_manager/moka_cache_manager.rs

```diff
@@ -0,0 +1,927 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::io::ErrorKind;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use async_walkdir::{Filtering, WalkDir};
+use base64::prelude::BASE64_URL_SAFE;
+use base64::Engine;
+use common_telemetry::{debug, warn};
+use futures::lock::Mutex;
+use futures::{FutureExt, StreamExt};
+use moka::future::Cache;
+use sha2::{Digest, Sha256};
+use snafu::ResultExt;
+use tokio::fs;
+use tokio_util::compat::{Compat, TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+use crate::error::{
+    CacheGetSnafu, CreateSnafu, MetadataSnafu, OpenSnafu, ReadSnafu, RemoveSnafu, RenameSnafu,
+    Result, RetriesExhaustedSnafu, WalkDirSnafu,
+};
+use crate::puffin_manager::cache_manager::{
+    BoxWriter, CacheManager, DirWriterProvider, InitBlobFn, InitDirFn,
+};
+use crate::puffin_manager::DirGuard;
+
+const MAX_RETRIES: usize = 3;
+
+const TMP_EXTENSION: &str = "tmp";
+const DELETED_EXTENSION: &str = "deleted";
+
+/// `MokaCacheManager` is a `CacheManager` that uses `moka` to manage cache.
+pub struct MokaCacheManager {
+    /// The base directory of the cache.
+    base_dir: PathBuf,
+
+    /// The cache maintaining the cache key to the size of the file or directory.
+    cache: Cache<String, CacheValue>,
+
+    /// The unreleased directories that are waiting to be released.
+    /// The value is the count of references to the directory.
+    unreleased_dirs: Arc<Mutex<HashMap<String, usize>>>,
```

Comment:
Is it possible to implement the `CacheValue` as a reference counted handle to a file or a dir? It might be similar to the [`FileHandle`](https://github.com/GreptimeTeam/greptimedb/blob/352cc9dddeeca9c57e4eb253934a6788a576969a/src/mito2/src/sst/file.rs#L139-L143) in mito.

One benefit is that if we `get` a cache value from the cache and keep holding the cache value, we can ensure the file/dir won't be removed.

We can move the file/dir path to the unreleased queue in `CacheValue::drop()` and release them later. What's more, since we don't track the count in `unreleased_dirs`, we can change the mutex to a sync one.

This also avoids the retry loop in `get_blob` as we can hold the returned `CacheValue`.

Reply:
Unlike `FileHandle` in mito, here the cache may perform multiple `get_blob`s on the same key. Consider a corner case:
1. The moka cache has already evicted a certain key, but the `FileHandle` for this key is still being held.
2. At this point, a new `get_blob` call arrives, and the file is initialized and overwritten.
3. The `FileHandle` held in step 1 is then dropped, which deletes the newly initialized file.
4. The new `get_blob` attempts to read the file and encounters a `NotFound` error.

Possible solutions to this problem are:
1. Have the `FileHandle` hold the file descriptor (fd), but this could lead to the potential problem of "opening too many files".
2. Have the cache track unreleased files, similar to how `unreleased_dirs` are handled. The way I handle `get_blob` now, the intention is essentially to obtain the fd. If that's achieved, it is considered successful. The cache is secondary and only needs to handle cleanup during eviction. Therefore, there's no need to track unreleased files.

Reply:
The handling approach for directories is with the intention of needing a `Mutex` to avoid potential unintended deletions. As mentioned in the previous comment, a reentrant `get_dir` will inevitably need to track unreleased directories.